docs: add migration guide (#16505)
sveltejs/kit: documentation/docs/60-appendix/35-migrating-to-sveltekit-3.md

---
title: Migrating to SvelteKit v3
---

SvelteKit 3 contains many small breaking changes, most of which you probably won't notice. The minimum dependency versions have been raised, a number of long-deprecated APIs have been removed, and configuration has moved from `svelte.config.js` into the Vite plugin. This guide lists every breaking change and how to migrate.

We highly recommend upgrading to the most recent 2.x version before upgrading to 3.0, and take advantage of its deprecating warnings before upgrading.

## Updated dependency requirements

SvelteKit 3 raises the minimum supported versions of its dependencies:

- **Node** 22 or newer
- **TypeScript** 6 or newer
- **Svelte** 5.56.4 or newer
- **Vite** 8.0.12 or newer (the first Vite 8 release bundling stable `rolldown` 1.0.0)
- **`@sveltejs/vite-plugin-svelte`** v7 or newer

Update the versions in your `package.json` and run your package manager's install command.

## Configuration is now passed to the Vite plugin

> `svelte.config.js` is no longer supported. SvelteKit configuration must now be passed to the `sveltekit()` Vite plugin in `vite.config.js`, and the `kit` prefix is dropped — options that used to live under `config.kit.*` are now top-level options of the plugin. They are places alongside other options such as `compilerOptions` for Svelte.

Reviewer: ```suggestion
`svelte.config.js` is no longer supported. SvelteKit configuration must now be passed to the `sveltekit()` Vite plugin in `vite.config.js`, and the `kit` prefix is dropped — options that used to live under `config.kit.*` are now top-level options of the plugin. They are placed alongside other options such as `compilerOptions` for Svelte.
```

## Changes to shallow routing

`pushState/replaceState` are deprecated in favor of `goto`:

```js
import { goto, pushState, replaceState } from '$app/navigation';

---pushState('/foo', state);---
+++goto('/foo', { shallow: true, state });+++

---replaceState('/bar', state);---
+++goto('/bar', { shallow: true, replace: true, state });+++
```

Shallow routing now triggers navigation hooks (`before/after/onNavigate`). You can filter them out by checking the `shallow` property of the object passed to those navigation hooks.

## External redirects must be opted into

`redirect(...)` to an external URL is now forbidden by default. To redirect to an external destination, pass an `external` option — either `true` to allow any external URL (except `javascript:` URLs, which are always blocked), or an array of allowed origins (through which you can allow `javascript:` URLs).

```js
import { redirect } from '@sveltejs/kit';

---throw redirect(307, 'https://example.com');---

> +++throw redirect(307, 'https://example.com', { external: true });+++

Reviewer 2: `redirect(...)` calls are thrown by SvelteKit
```suggestion
---redirect(307, 'https://example.com');---
+++redirect(307, 'https://example.com', { external: true });+++
```

- **`adapter-node`** — bundled with `rolldown`; the `ORIGIN` environment variable is removed (use `paths.origin` in config).
- **`adapter-netlify`** — output now conforms to the stable [Netlify Frameworks API](https://docs.netlify.com/build/frameworks/frameworks-api/); deploying/previewing with the Netlify CLI requires `v17.31.0` or later (`npm i -g netlify-cli@latest`); edge function build target is `es2022`.
- **`adapter-vercel`** — edge function build target is now `es2022`; edge functions are bundled with `rolldown`.

## Routing and project structure

### Consistent special filename patterns

Special filenames are now matched on a _segment_ basis rather than only via the `*.server.*` / `*.remote.*` infix. A `server` segment anywhere in a file's path makes it server-only, and a `remote` segment anywhere in the path makes it a remote module.

This means `*.remote.*.ts` now works the same way `*.server.*.ts` always has, and bare files named `server.ts` or `remote.ts` are now treated as special (previously they were not). If you have a file named `server.ts` or `remote.ts` that should _not_ be treated as server-only/remote, rename it.

### Server-only directories

> Any directory named `server` in the path is now treated as server-only everywhere in the project, not just `$lib/server`; with the exception of `src/routes` and the assets directory.

Reviewer 3: Should we word this differently, since we're getting rid of `$lib`? What about just `src/lib/server`?

Reviewer 4: ```suggestion
Any directory named `server` in the path is now treated as server-only everywhere in the project, not just `src/lib/server`; with the exception of `src/routes` and the assets directory.
```

If you have a `server/` directory that is meant to be importable from client code, rename it. Otherwise this change simply expands protection against accidentally importing server code into the browser.

### Universal `config` takes precedence over server `config`

`config` exported from a universal `+page.js`/`+layout.js` now takes precedence over `config` exported from the corresponding `+page.server.js`/`+layout.server.js`, matching how other page options are resolved. If you export `config` from both, move the canonical export to the universal file or consolidate them.

## Modules and imports

> ### `@sveltejs/kit/node/polyfills` has been removed

Reviewer 5: The guide doesn't mention `$service-worker`, which was deleted in #16450.

````suggestion
### `$service-worker` has been removed

Use `immutable`, `assets` and `prerendered` from `$app/manifest` in place of `build`, `files` and `prerendered`. `version` now comes from `$app/env`, and `resolve(...)` from `$app/paths` replaces `base`.

```js
---import { build, files, version } from '$service-worker';---
+++import { immutable, assets } from '$app/manifest';
import { version } from '$app/env';+++
```

### `@sveltejs/kit/node/polyfills` has been removed
````

### `csrf.checkOrigin` replaced by `csrf.trustedOrigins`

The deprecated `csrf.checkOrigin` option has been removed. CSRF protection is always on; instead of disabling it with `checkOrigin: false`, allowlist trusted cross-origin hosts with `csrf.trustedOrigins`.

```js
csrf: {
	---checkOrigin: false---
	+++trustedOrigins: ['https://trusted-site.com']+++
}
```

### Cross-origin form submissions require a `Content-Type` header

Cross-origin form submissions that omit a `Content-Type` header are now rejected as CSRF, where previously they were allowed through ([#16347](https://github.com/sveltejs/kit/pull/16347)). Ensure cross-origin form submissions include a `Content-Type` header, or add the origin to `csrf.trustedOrigins`.

### CORS for static assets in development is handled by Vite

SvelteKit no longer sets `access-control-allow-origin: *` on every static asset request in development. CORS is now delegated to Vite's built-in middleware. If you rely on cross-origin access to static assets in dev, configure it in your Vite config:

```js
export default defineConfig({
	server: {
+++		cors: { origin: '*' }+++
	}
});
```

## Remote functions

### Remote functions require an opt-in

> Files matching `*.remote.ts`/`*.remote.js` now error during development and builds unless `experimental.remoteFunctions` is enabled. As such they are now reserved for remote functions.

Reviewer 6: Would this be more accurate?

```suggestion
Files matching `*remote.ts`/`*remote.js` now error during development and builds unless `experimental.remoteFunctions` is enabled. As such they are now reserved for remote functions.
```

Reviewer 4: No but this would

```suggestion
Files with a `remote` segment in the name now error during development and builds unless `experimental.remoteFunctions` is enabled. As such they are now reserved for remote functions.
```

## Updated dependency requirements

SvelteKit 3 requires the following minimum versions:

- Node v22.17
- TypeScript v6
- Svelte v5.56.4
- Vite v8.0.12 (the first Vite 8 release bundling stable `rolldown` v1)
- `@sveltejs/vite-plugin-svelte` v7

Update the versions in your `package.json` and run your package manager's install command.

## Configuration

### svelte.config.js is no longer supported

Instead of declaring project configuration in `svelte.config.js`, it must now be passed to the `sveltekit` Vite plugin in `vite.config.js`. Options that previously lived under `config.kit.*` are now top-level plugin options, alongside things like `compilerOptions`:

```js
/// file: vite.config.js
import { defineConfig } from 'vite';
import { sveltekit } from '@sveltejs/kit/vite';
import adapter from '@sveltejs/adapter-auto';

export default defineConfig({
	plugins: [
+++		sveltekit({
			compilerOptions: { experimental: { async: true }},
			adapter: adapter()
		})+++
	]
});
```

See the [configuration docs](configuration) for further examples.

### Removed options

The following options are obsolete and should be removed from your `vite.config.js`:

> - `files.lib` (see [`$lib is now `#lib`](#$lib-is-now-lib))

Reviewer 6: Little formatting typo

```suggestion
- `files.lib` (see [`$lib` is now `#lib`](#$lib-is-now-lib))
```

### Errors are typed as `App.Error | undefined`

The `error` property on remote function resources (queries, live queries, forms, prerender functions) is now typed as `App.Error | undefined` rather than `any`, as the error is always transformed by `handleError`.

### Form submissions require `field.as(...)`

A form control must use attributes from a field associated with the current `form` object:

```svelte
<input {...myform.fields.message.as('text')}>
```

Manually specifying a name (`<input name="message">`) will cause the submission to be rejected.

## Miscellaneous

### Links to the current page cause a refresh

If the user clicks a link that points to the current location, SvelteKit will `refreshAll()` instead of doing nothing.

### `data-sveltekit-*` uses `false` instead of `'off'`

The `'off'` value for `data-sveltekit-*` link attributes has been removed in favour of `false`.

```svelte
---<a href="..." data-sveltekit-preload-data="off">---

> +++<a href="..." data-sveltekit-preload-data="false">+++

Author: ```suggestion
+++<a href="..." data-sveltekit-preload-data={false}>+++
```

Reviewer 4: this wouldn't work in an `app.html`, whereas `"false"` works everywhere

### Errors are typed as `App.Error | undefined`

The `error` property on remote function resources (queries, live queries, forms, prerender functions) is now typed as `App.Error | undefined` rather than `any`, as the error is always transformed by `handleError`.

### Form submissions require `field.as(...)`

A form control must use attributes from a field associated with the current `form` object:

```svelte
<input {...myform.fields.message.as('text')}>
```

Manually specifying a name (`<input name="message">`) will cause the submission to be rejected.

## Miscellaneous

### Links to the current page cause a refresh

If the user clicks a link that points to the current location, SvelteKit will `refreshAll()` instead of doing nothing.

### `data-sveltekit-*` uses `false` instead of `'off'`

The `'off'` value for `data-sveltekit-*` link attributes has been removed in favour of `false`.

```svelte
---<a href="..." data-sveltekit-preload-data="off">---

> +++<a href="..." data-sveltekit-preload-data={false}>+++

Author: ```suggestion
+++<a href="..." data-sveltekit-preload-data="false">+++
```
